runtime/src/chpl-qsbr.c

```diff
@@ -0,0 +1,226 @@
+/*
+ * Copyright 2004-2018 Cray Inc.
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "chplrt.h"
+#include "chpl-thread-local-storage.h"
+#include "chpl-mem.h"
+#include "chpl-atomics.h"
+#include "chpl-tasks.h"
+#include "chpl-qsbr.h"
+
+static uint64_t global_epoch = 0;
+
+struct defer_data {
+    void *data;
+    int numData;
+};
+
+// Deferred deletion list and associated epoch
+// Ideas: First thread that does not complete within a
+// reasonable amount of time gets to handle deleting it.
+// If a thread hits a checkpoint and seeing a deferred node,
+// it will process one of them; if it is unable to, it will
+// 'pass the buck' to the next problematic thread ensuring that
+// eventually it will be deleted.
+struct defer_node {
+  uint64_t targetEpoch;
+  struct defer_data ddata;
+  struct defer_node *next;
+};
+
+// Thread-local data that keeps track of our observed global state.
+struct tls_node {
+  /*
+    The current observed epoch. This is used to ensure that writers
+    do not delete an older instance that corresponds to this epoch.
+  */
+  volatile uint64_t epoch;
```

Comment:
I generally view `volatile` as a warning sign in multi-thread code. Might it be more reasonable to make it an atomic, and do "relaxed" reads?

Reply:
I'd like to see these use atomic_uintptr_t / atomic_uint_least64_t to emphasize that they need atomic read/write.